packages/block-editor/src/components/block-list/block.js

```diff
@@ -544,7 +546,15 @@ function BlockListBlock( {
 						ref={ breadcrumb }
 					/>
 				) }
-				{ ( shouldShowContextualToolbar || isForcingContextualToolbar.current ) && (
+				{ ( ( ! consumeChildToolbar && shouldShowContextualToolbar ) || isForcingContextualToolbar.current ) && (
+					<BlockContextualToolbar
+						// If the toolbar is being shown because of being forced
+						// it should focus the toolbar right after the mount.
+						focusOnMount={ isForcingContextualToolbar.current }
+					/>
+				) }
+
+				{ isParentOfSelectedBlock && isRootOfHierarchy && (
 					<BlockContextualToolbar
```

Comment:
BlockContextualToolbar renders the toolbars elements added by slot&fill but also things like the block switcher and more menu options etc.
This mechanism is working well for things added by the block that pass by the slot & fill, but when clicks happen outside on the more menu, for example, the parent block gets selected (as you pointed out on the core editor chat).
It is not exclusive of the more menu e.g: if we click on child block menu icon the parent gets selected, it is also impossible to use the block switcher (possible to test by changing the allowed blocks of navigation menu to allowedBlocks={ [ 'core/navigation-menu-item', 'core/group' ] } so the group transform is available).
This happens because the click event is happening on the dom of the parent. When a click occurs on the dom of a block, that block gets selected.

I think we have three solutions:
- Change the code that selects a block to be aware of this logic and do nothing if the toolbar being shown is one of the children. The code will need to check if the event is happening specifically inside the toolbar as if the click occurs outside the toolbar, the parent should get selected.
- Use IgnoreNestedEvents on the tollbar, so events do not propagate to the code that selects the block. I'm not sure if this has unintended consequences and causes regressions I am not aware of.
- Use a slot & fill mechanism. Blocks render a slot for the child toolbar. The child BlockContextualToolbar is rendered inside the fill of the parent if the flag is true. This solution, although following a different path, may solve the problem of multiple nesting levels as in slot & fill fills are rendered on the closest ancestor slot.

Reply:
> This mechanism is working well for things added by the block that pass by the slot & fill, but when clicks happen outside on the more menu

> This happens because the click event is happening on the dom of the parent. When a click occurs on the dom of a block, that block gets selected.

You are right. This is what [I observed](https://github.com/WordPress/gutenberg/pull/18440#issuecomment-553396286).

> Change the code that selects a block to be aware of this logic and do nothing if the toolbar being shown is one of the children. 

I considered this. This would need to rely on particular DOM characteristic to detect whether the event occured within a toolbar. Not sure how robust it would be.

> Use IgnoreNestedEvents on the tollbar, so events do not propagate to the code that selects the block.

I considered this also. Was trying to grok `IgnoreNestedEvents` yesterday. May need some more advice on it.

> Use a slot & fill mechanism.

I'm in favour of exploring this as it seems to solve an number of the pain points. Need to get my head around how this might work.

Reply:
I've gone with the Slot/Fill approach and I can confirm that the entire toolbar is now usable as the events are set to bubble within the React component tree rather than the DOM tree. This means the events never selects the parent Block.

![Screen Capture on 2019-11-15 at 13-19-28](https://user-images.githubusercontent.com/444434/68946353-b1614c80-07aa-11ea-8bbb-3581915ab4f7.gif)